InnerEye/ML/SSL/datamodules_and_datasets/cxr_datasets.py

```diff
@@ -0,0 +1,155 @@
+        self._prepare_dataset()
+
+    def _prepare_dataset(self) -> None:
+        self.indices: List[int] = []
+        self.filenames: List[PathOrString] = []
+        raise NotImplementedError("_prepare_dataset needs to be implemented by the child classes.")
+
+    def __getitem__(self, index: int) -> OptionalIndexInputAndLabel:
+        """
+        :param index: The index of the sample to be fetched
+        :return: The image and (fake) label tensors
+        """
+        filename = self.filenames[index]
+        target = self.targets[index] if self.targets is not None else 0
+        if is_dicom_file_path(str(filename)):
+            scan_image = load_dicom_image(filename)
+            # Dicom files have arbitrary pixel intensities, convert to [0,255] range so that PIL can
+            # read the array into an Image object.
+            scan_image = (scan_image - scan_image.min()) * 255. / (scan_image.max() - scan_image.min())
+            # Load as PIL Image in grey-scale (convert("L") step), yields a 1-channel image.
+            scan_image = Image.fromarray(scan_image).convert("L")
+        else:
+            # Load as PIL Image in grey-scale (convert("L") step), yields a 1-channel image with pixel values in range
+            # [0,1] (float).
+            scan_image = Image.open(filename).convert("L")
+        if self.transforms is not None:
+            scan_image = self.transforms(scan_image)
+        return scan_image, target
+
+    def __len__(self) -> int:
+        """
+        :return: The size of the dataset
+        """
+        return len(self.indices)
+
+
+class InnerEyeCXRDatasetWithReturnIndex(InnerEyeDataClassBaseWithReturnIndex, InnerEyeCXRDatasetBase):
+    """
+    Any dataset used in SSL needs to inherit from InnerEyeDataClassBaseWithReturnIndex as well as VisionData.
+    This class is just a shorthand notation for this double inheritance.
+    """
+    pass
+
+
+class RSNAKaggleCXR(InnerEyeCXRDatasetWithReturnIndex):
+    """
+    Dataset class to load the RSNA Chest-Xray training dataset. Use all the data for train and val. No test data
+    implemented.
+    """
+
+    def _prepare_dataset(self) -> None:
+        if self.train:
+            self.dataset_dataframe = pd.read_csv(self.root / "dataset.csv")
+            self.targets = self.dataset_dataframe.label.values.astype(np.int64)
+            self.subject_ids = self.dataset_dataframe.subject.values
+            self.indices = np.arange(len(self.dataset_dataframe))
+            self.filenames = [self.root / f"{subject_id}.dcm" for subject_id in self.subject_ids]
+        else:
+            # No test set implemented for this data class.
+            self.indices = []
+            self.filenames = []
+
+    @property
+    def num_classes(self) -> int:
+        return 2
+
+
+class NIH(InnerEyeCXRDatasetWithReturnIndex):
+    """
+    Dataset class to load the NIH Chest-Xray dataset. Use the full data for training and validation (including
```

Comment:
Same in here - I just saw that it's included in the markdown file - maybe we can just add a cross-link to the file:
docs/self_supervised_models.md